subprojects/core/src/main/java/org/gradle/internal/service/scopes/ToolingServices.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service.scopes;
+
+import org.gradle.internal.service.ServiceRegistration;
+
+public class ToolingServices extends AbstractPluginServiceRegistry {
+
+    @Override
+    public void registerBuildServices(ServiceRegistration registration) {
+        registration.addProvider(new ExceptionServices());
+    }
+
+    static class ExceptionServices {
+        public ExceptionCollector createExceptionCollector() {
+            String classpathMode = System.getProperty("org.gradle.kotlin.dsl.provider.mode");
```

Comment:
@eskatos This code belongs to the kotlin-dsl submodule, but it might not be worth the complexity to extract the code.